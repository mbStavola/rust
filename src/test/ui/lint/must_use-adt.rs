#![deny(unused_must_use)]

#[must_use]
struct S;

#[must_use]
trait A {}

struct B;

impl A for B {}

struct T(S);

struct U {
    x: (),
    y: T,
}

struct V {
    a: S,
}

struct W {
    w: [(u8, Box<dyn A>); 2],
    x: u32,
    y: (B, B),
    z: (S, S),
    e: [(u8, Box<dyn A>); 2],
    f: S,
}

fn get_v() -> V {
    V { a: S }
}

struct Z([(u8, Box<dyn A>); 2]);

fn get_wrapped_arr() -> Z {
    Z([(0, Box::new(B)), (0, Box::new(B))])
}

fn get_tuple_arr() -> ([(u8, Box<dyn A>); 2],) {
    ([(0, Box::new(B)), (0, Box::new(B))],)
}

struct R<T> {
    r: T
}

struct List<T>(T, Option<Box<Self>>);

fn main() {
    S; //~ ERROR unused `S` that must be used
    T(S); //~ ERROR unused `S` in field `0` that must be used
    U { x: (), y: T(S) }; //~ ERROR unused `S` in field `0` that must be used
    get_v(); //~ ERROR unused `S` in field `a` that must be used
    V { a: S }; //~ ERROR unused `S` in field `a` that must be used
    W {
        w: [(0, Box::new(B)), (0, Box::new(B))],
        //~^ ERROR unused array of boxed `A` trait objects in tuple element 1 that must be used
        x: 0,
        y: (B, B),
        z: (S, S),
        //~^ unused `S` in tuple element 0 that must be used
        //~^^ unused `S` in tuple element 1 that must be used
        e: [(0, Box::new(B)), (0, Box::new(B))],
        //~^ unused array of boxed `A` trait objects in tuple element 1 that must be used
        f: S, //~ ERROR unused `S` in field `f` that must be used
    };
    get_wrapped_arr();
    //~^ ERROR unused array of boxed `A` trait objects in tuple element 1 that must be use
    get_tuple_arr();
    //~^ ERROR unused array of boxed `A` trait objects in tuple element 1 that must be used
    R { r: S }; //~ ERROR unused `S` in field `r` that must be used
    List(S, Some(Box::new(List(S, None)))); //~ ERROR unused `S` in field `0` that must be used
}
