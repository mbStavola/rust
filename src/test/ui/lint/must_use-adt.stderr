error: unused `S` that must be used
  --> $DIR/must_use-adt.rs:54:5
   |
LL |     S;
   |     ^^
   |
note: lint level defined here
  --> $DIR/must_use-adt.rs:1:9
   |
LL | #![deny(unused_must_use)]
   |         ^^^^^^^^^^^^^^^

error: unused `S` in field `0` that must be used
  --> $DIR/must_use-adt.rs:55:7
   |
LL |     T(S);
   |       ^

error: unused `S` in field `0` that must be used
  --> $DIR/must_use-adt.rs:56:21
   |
LL |     U { x: (), y: T(S) };
   |                     ^

error: unused `S` in field `a` that must be used
  --> $DIR/must_use-adt.rs:57:5
   |
LL |     get_v();
   |     ^^^^^^^^

error: unused `S` in field `a` that must be used
  --> $DIR/must_use-adt.rs:58:12
   |
LL |     V { a: S };
   |            ^

error: unused array of boxed `A` trait objects in tuple element 1 that must be used
  --> $DIR/must_use-adt.rs:60:12
   |
LL |         w: [(0, Box::new(B)), (0, Box::new(B))],
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused `S` in tuple element 0 that must be used
  --> $DIR/must_use-adt.rs:64:13
   |
LL |         z: (S, S),
   |             ^

error: unused `S` in tuple element 1 that must be used
  --> $DIR/must_use-adt.rs:64:16
   |
LL |         z: (S, S),
   |                ^

error: unused array of boxed `A` trait objects in tuple element 1 that must be used
  --> $DIR/must_use-adt.rs:67:12
   |
LL |         e: [(0, Box::new(B)), (0, Box::new(B))],
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused `S` in field `f` that must be used
  --> $DIR/must_use-adt.rs:69:12
   |
LL |         f: S,
   |            ^

error: unused array of boxed `A` trait objects in tuple element 1 that must be used
  --> $DIR/must_use-adt.rs:71:5
   |
LL |     get_wrapped_arr();
   |     ^^^^^^^^^^^^^^^^^^

error: unused array of boxed `A` trait objects in tuple element 1 that must be used
  --> $DIR/must_use-adt.rs:73:5
   |
LL |     get_tuple_arr();
   |     ^^^^^^^^^^^^^^^^

error: unused `S` in field `r` that must be used
  --> $DIR/must_use-adt.rs:75:12
   |
LL |     R { r: S };
   |            ^

error: unused `S` in field `0` that must be used
  --> $DIR/must_use-adt.rs:76:10
   |
LL |     List(S, Some(Box::new(List(S, None))));
   |          ^

error: aborting due to 14 previous errors

